### PR TITLE
3137 UI refresh token logic

### LINKED ISSUE #3137: Generate refresh token
## Generate our own refresh token:
We could generate a token that's exactly the same as the authentication token, but with an extended lifespan (we need to use a custom TimestampSigner that uses not the current time, but current time + REFRESH_TIMEDELTA as its timestamp). The only difference between the authentication token and the refresh token would be that the refresh token has a longer lifespan. Then, we would add another endpoint.

## Tasks:

 - [x] Create a refresh token (0d) @VakarisZ 
 - [x] Modify login and registration to return refresh token (0d) @VakarisZ 
 - [x] Add an endpoint that allows you to login with a refresh token
 - [x] Make a refresh token distinct from access token @VakarisZ 
 - [x] Add logic to the agent to use the refresh token (0d) - @shreyamalviya 
 - [x] Adjust the times for refresh and authentication token (0d) @VakarisZ 
 - [x] Redesign refresh strategy (1.25d)
    - [x] BB test to verify old tokens are invalid after refresh
    - [x] Add expiration time to all responses with tokens
    - [x] Modify the agent to use the new scheme
    - [x] Make the agent's HTTPIslandAPIClient to be multiprocessing-safe @cakekoa 
    - [x] Add logic to the UI to use the refresh token @ilija-lazoroski 
 - [ ] Cleanup vulture allowlist

### PR DESCRIPTION
**Pull request by ilija-lazoroski**
# What does this PR do?

Fixes #3137 .

Add any further explanations here.

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [x] Is the TravisCI build passing?
* [x] Was the CHANGELOG.md updated to reflect the changes?
* [x] Was the documentation framework updated to reflect the changes?
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Do all unit tests pass?
* [ ] Do all end-to-end tests pass?
* [x] Any other testing performed?
    > Tested by {Running the Monkey locally with relevant config/running Island/...}
* [x] If applicable, add screenshots or log transcripts of the feature working

### CODE REVIEW

File: monkey/monkey_island/cc/ui/src/services/AuthService.js [hunk 1]
```diff
@@ -81,8 +118,8 @@ export default class AuthService {
     })
   };
 
-  _getTokenFromResponse= (response) => {
-    return _.get(response, 'response.user.'+this.TOKEN_NAME_IN_RESPONSE, undefined);
+  _getTokenFromResponse= (tokenName, response) => {
+    return _.get(response, 'response.user.'+tokenName, undefined);
```
Comment (ordabach) [inline monkey/monkey_island/cc/ui/src/services/AuthService.js:122]: Instead of lodash use optional chaining (?.):
`return response?.user?.[tokenName];`

You can read more about it [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining)

File: monkey/monkey_island/cc/ui/src/services/AuthService.js [hunk 2]
```diff
@@ -29,16 +34,45 @@ export default class AuthService {
...
+    this._refreshAuthToken(url, options);
     return this._authFetch(url, options);
   };
 
+  _refreshAuthToken = (url, options) => {
+    this._authFetch(url, options).then(res => {
+      if(res.status == 401)
+      {
+        let refreshToken = this._getToken(this.REFRESH_TOKEN_NAME_IN_LOCALSTORAGE);
+        if (refreshToken !== undefined && refreshToken !== null) {
```
Comment (ordabach) [inline monkey/monkey_island/cc/ui/src/services/AuthService.js:53]: Same as writing `if(refreshToken) { // rest of the code }`
I think it will be more understandable (it will return `false` if the 'refreshToken' value is null or undefined.

File: monkey/monkey_island/cc/ui/src/services/AuthService.js [hunk 3]
```diff
@@ -29,16 +33,47 @@ export default class AuthService {
...
       });
   }
 
   authFetch = (url, options) => {
+    this._refreshAuthToken(url, options);
     return this._authFetch(url, options);
   };
 
+  _refreshAuthToken = (url, options) => {
+    this._authFetch(url, options).then(res => {
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/ui/src/services/AuthService.js:48]: After a new token has been acquired, shouldn't this attempt the original request again? Maybe I'm just missing something?

Comment (ilija-lazoroski) [reply monkey/monkey_island/cc/ui/src/services/AuthService.js:48]: See where this is used and what it is returned after it

Comment (mssalvatore) [reply monkey/monkey_island/cc/ui/src/services/AuthService.js:48]: So every call is made twice?

```javascript
  authFetch = (url, options) => {
    this._refreshAuthToken(url, options);
    return this._authFetch(url, options);
  };

  _refreshAuthToken = (url, options) => {
    this._authFetch(url, options).then(res => {
      if(res.status == 401)
      {
        let refreshToken = this._getToken(this.REFRESH_TOKEN_NAME_IN_LOCALSTORAGE);
        if (refreshToken) {
        this._fetchTokenPairFromRefreshToken(this.REFRESH_AUTH_TOKEN_ENDPOINT, {
          method: 'POST',
          body: JSON.stringify({
...
```

File: monkey/monkey_island/cc/ui/src/services/AuthService.js [hunk 4]
```diff
@@ -29,16 +33,47 @@ export default class AuthService {
...
+          })
+        }).then(response => response.json().then(data => ({status: response.status, body: data})))
+          .then(object => {
+            if(object.status === 200) {
+              let authToken = this._getTokenFromResponse(this.AUTH_TOKEN_NAME_IN_RESPONSE, object.body);
+              this._setToken(this.AUTH_TOKEN_NAME_IN_LOCALSTORAGE, authToken);
+
+              let refreshToken = this._getTokenFromResponse(this.REFRESH_TOKEN_NAME_IN_RESPONSE, object.body);
+              console.log(authToken, refreshToken);
+              this._setToken(this.REFRRESH_TOKEN_NAME_IN_LOCALSTORAGE, refreshToken);
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/ui/src/services/AuthService.js:66]: How did you test this?
```suggestion
              this._setToken(this.REFRESH_TOKEN_NAME_IN_LOCALSTORAGE, refreshToken);
```

File: monkey/monkey_island/cc/ui/src/services/AuthService.js [hunk 5]
```diff
@@ -29,16 +33,47 @@ export default class AuthService {
...
+            [this.REFRESH_TOKEN_NAME_IN_RESPONSE]: this._getToken(this.REFRESH_TOKEN_NAME_IN_LOCALSTORAGE)
+          })
+        }).then(response => response.json().then(data => ({status: response.status, body: data})))
+          .then(object => {
+            if(object.status === 200) {
+              let authToken = this._getTokenFromResponse(this.AUTH_TOKEN_NAME_IN_RESPONSE, object.body);
+              this._setToken(this.AUTH_TOKEN_NAME_IN_LOCALSTORAGE, authToken);
+
+              let refreshToken = this._getTokenFromResponse(this.REFRESH_TOKEN_NAME_IN_RESPONSE, object.body);
+              console.log(authToken, refreshToken);
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/ui/src/services/AuthService.js:65]: Don't forget to remove this.

File: monkey/monkey_island/cc/ui/src/services/AuthService.js [hunk 6]
```diff
@@ -29,16 +33,47 @@ export default class AuthService {
...
         if(response.meta.code === 200){
-          this._removeToken();
+          this._removeToken(this.AUTH_TOKEN_NAME_IN_LOCALSTORAGE);
         }
         return response;
       });
   }
 
   authFetch = (url, options) => {
+    this._refreshAuthToken(url, options);
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/ui/src/services/AuthService.js:43]: We're refreshing the token on every request?

File: monkey/monkey_island/cc/ui/src/services/AuthService.js [hunk 7]
```diff
@@ -109,12 +147,32 @@ export default class AuthService {
           res.clone().json().then(res_json => {
             console.log('Got 401 from server while trying to authFetch: ' + JSON.stringify(res_json));
           });
-          this._removeToken();
+          this._removeToken(this.AUTH_TOKEN_NAME_IN_LOCALSTORAGE);
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/ui/src/services/AuthService.js:150]: The token is being removed here. What happens when line 68 is executed then?

File: monkey/monkey_island/cc/ui/src/services/AuthService.js [hunk 8]
```diff
@@ -109,12 +147,32 @@ export default class AuthService {
...
+      'Content-Type': 'application/json'
+    };
+
+    if (Object.prototype.hasOwnProperty.call(options, 'headers')) {
+      for (let header in headers) {
+        options['headers'][header] = headers[header];
+      }
+    } else {
+      options['headers'] = headers;
+    }
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/ui/src/services/AuthService.js:168]: Why is this needed?

File: monkey/monkey_island/cc/ui/src/services/AuthService.js [hunk 9]
```diff
@@ -29,16 +33,47 @@ export default class AuthService {
...
+
+              let refreshToken = this._getTokenFromResponse(this.REFRESH_TOKEN_NAME_IN_RESPONSE, object.body);
+              console.log(authToken, refreshToken);
+              this._setToken(this.REFRRESH_TOKEN_NAME_IN_LOCALSTORAGE, refreshToken);
+            } else if (object.status === 401) {
+              this._removeToken(this.AUTH_TOKEN_NAME_IN_LOCALSTORAGE);
+              this._removeToken(this.REFRESH_TOKEN_NAME_IN_LOCALSTORAGE);
+            }
+          })
+        }
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/ui/src/services/AuthService.js:72]: This should be cleaned up. This function is too big and hard to read.

File: monkey/monkey_island/cc/ui/src/services/AuthService.js (hunk 3 above)
Comment (shreyamalviya) [reply monkey/monkey_island/cc/ui/src/services/AuthService.js:48]: Why are you calling `_authFetch()` inside `_refreshAuthToken()`? Just check the response status code in `authFetch()`, refresh the token if required, and make the request again.

**Review by shreyamalviya — CHANGES_REQUESTED** [review]

File: monkey/monkey_island/cc/ui/src/services/AuthService.js (hunk 6 above)
Comment (ilija-lazoroski) [reply monkey/monkey_island/cc/ui/src/services/AuthService.js:43]: A mistake which will be fixed.

File: monkey/monkey_island/cc/ui/src/services/AuthService.js (hunk 7 above)
Comment (ilija-lazoroski) [reply monkey/monkey_island/cc/ui/src/services/AuthService.js:150]: Nothing much, two different things. Ff for some reason we have the auth token still in the storage then we need to remove it on 401 from the `refresh-authentication-token` endpoint. The one on 150 is if we have 401 on whatever other request.

File: monkey/monkey_island/cc/ui/src/services/AuthService.js (hunk 8 above)
Comment (ilija-lazoroski) [reply monkey/monkey_island/cc/ui/src/services/AuthService.js:168]: To set up the options which may go away.

File: monkey/monkey_island/cc/ui/src/services/AuthService.js (hunk 4 above)
Comment (ilija-lazoroski) [reply monkey/monkey_island/cc/ui/src/services/AuthService.js:66]: Lowered down the expiration time. And waited a couple of times but I messed something up I guess. I will check

File: monkey/monkey_island/cc/ui/src/services/AuthService.js (hunk 7 above)
Comment (shreyamalviya) [reply monkey/monkey_island/cc/ui/src/services/AuthService.js:163]: > for some reason we have the auth token still in the storage then we need to remove it on 401 from the refresh-authentication-token endpoint.

What happens when it's already removed (because of line 150)? Does it crash?

File: monkey/monkey_island/cc/ui/src/services/AuthService.js [hunk 10]
```diff
@@ -29,16 +33,60 @@ export default class AuthService {
...
+          }
+        })
+    }
+  }
+
+  _fetchTokenPairFromRefreshToken() {
+    const options = {
+      method: 'POST',
+      body: JSON.stringify({
+        [this.REFRESH_TOKEN_NAME_IN_RESPONSE]: this._getToken(this.REFRESH_TOKEN_NAME_IN_LOCALSTORAGE)
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/ui/src/services/AuthService.js:72]: Why is this in square brackets?

File: monkey/monkey_island/cc/ui/src/services/AuthService.js (hunk 7 above)
Comment (ordabach) [reply monkey/monkey_island/cc/ui/src/services/AuthService.js:163]: If the token doesn't exist it won't do anything

Comment (ilija-lazoroski) [reply monkey/monkey_island/cc/ui/src/services/AuthService.js:163]: Nope, the localstorage can't find it so nothing to delete.

File: monkey/monkey_island/cc/ui/src/services/AuthService.js [hunk 11]
```diff
@@ -48,12 +96,15 @@ export default class AuthService {
...
     }).then(response => response.json())
       .then(res => {
-        let token = this._getTokenFromResponse(res);
-        if (token !== undefined) {
-          this._setToken(token);
+        let authToken = this._getTokenFromResponse(this.AUTH_TOKEN_NAME_IN_RESPONSE, res);
+        let refreshToken = this._getTokenFromResponse(this.REFRESH_TOKEN_NAME_IN_RESPONSE, res);
+        if (authToken !== undefined && refreshToken !== undefined) {
+          this._setToken(this.AUTH_TOKEN_NAME_IN_LOCALSTORAGE, authToken);
+          this._setToken(this.REFRESH_TOKEN_NAME_IN_LOCALSTORAGE, refreshToken);
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/ui/src/services/AuthService.js:103]: This is duplicated in a couple of places. Let's add something like `setTokensInLocalStorage(authToken, refreshToken)`.

File: monkey/monkey_island/cc/ui/src/services/AuthService.js [hunk 12]
```diff
@@ -48,12 +96,15 @@ export default class AuthService {
...
           return {result: true};
         } else {
-          this._removeToken();
+          this._removeToken(this.AUTH_TOKEN_NAME_IN_LOCALSTORAGE);
+          this._removeToken(this.REFRESH_TOKEN_NAME_IN_LOCALSTORAGE);
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/ui/src/services/AuthService.js:107]: Same comment as above

**Review by shreyamalviya — CHANGES_REQUESTED** [review]

File: monkey/monkey_island/cc/ui/src/services/AuthService.js (hunk 10 above)
Comment (ilija-lazoroski) [reply monkey/monkey_island/cc/ui/src/services/AuthService.js:72]: I can't find the link but it needs square brackets so it will take the value of the variable when sending the request.

Comment (ordabach) [reply monkey/monkey_island/cc/ui/src/services/AuthService.js:75]: @shreyamalviya that's the correct syntax for creating [computed property names](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer#computed_property_names)

**Review by shreyamalviya — APPROVED** [review]

**Review by ordabach — APPROVED** [review]

File: monkey/monkey_island/cc/ui/src/services/AuthService.js [hunk 13]
```diff
@@ -29,16 +33,62 @@ export default class AuthService {
       .then(response => response.json())
       .then(response => {
         if(response.meta.code === 200){
-          this._removeToken();
+          this._removeTokens([this.AUTH_TOKEN_NAME_IN_LOCALSTORAGE]);
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/ui/src/services/AuthService.js:36]: Why are we not also removing the refresh token?

**Review by mssalvatore — CHANGES_REQUESTED** [review]

File: monkey/monkey_island/cc/ui/src/services/AuthService.js [hunk 14]
```diff
@@ -29,16 +33,62 @@ export default class AuthService {
...
+        return this._authFetch(url, options);
+      }
+      return res;
+    });
   };
 
+  _refreshAuthToken = () => {
+    let refreshToken = this._getToken(this.REFRESH_TOKEN_NAME_IN_LOCALSTORAGE);
+    if (refreshToken) {
+      this._fetchTokenPairFromRefreshToken()
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/ui/src/services/AuthService.js:55]: The name of this is confusing. It makes me think you're going to parse the refresh token and derive a new token pair from it.
```suggestion
      this._fetchNewTokenPair()
```

File: monkey/monkey_island/cc/ui/src/services/AuthService.js [hunk 15]
```diff
@@ -125,20 +179,23 @@ export default class AuthService {
...
+  _setTokenPair(authToken, refreshToken) {
+    localStorage.setItem(this.AUTH_TOKEN_NAME_IN_LOCALSTORAGE, authToken);
+    localStorage.setItem(this.REFRESH_TOKEN_NAME_IN_LOCALSTORAGE, refreshToken);
   }
 
-  _removeToken() {
-    localStorage.removeItem(this.TOKEN_NAME_IN_LOCALSTORAGE);
+  _removeTokens(tokenNameArray) {
+      tokenNameArray?.forEach(tokenName => {
+        localStorage.removeItem(tokenName);
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/ui/src/services/AuthService.js:193]: Is there any reason we don't want to always remove the auth and refresh tokens together?

Comment (ilija-lazoroski) [reply monkey/monkey_island/cc/ui/src/services/AuthService.js:193]: Yep, there is. When we do a request and it is 401, we can immediately remove the authentication token but not the refresh token as we need it so we can generate a new token pair.

**Comment by mssalvatore** [comment]
We went a different direction.
